common/src/main/java/io/netty/util/concurrent/DefaultFutureCompletionStage.java

```diff
@@ -0,0 +1,563 @@
+    public <U> FutureCompletionStage<U> applyToEitherAsync(
+            CompletionStage<? extends V> other, Function<? super V, U> fn) {
+        return applyToEitherAsync(other, fn, ForkJoinPool.commonPool());
+    }
+
+    @Override
+    public FutureCompletionStage<Void> acceptEither(CompletionStage<? extends V> other, Consumer<? super V> action) {
+        return acceptEitherAsync(other, action, SAME_AS_FUTURE);
+    }
+
+    @Override
+    public FutureCompletionStage<Void> acceptEitherAsync(
+            CompletionStage<? extends V> other, Consumer<? super V> action) {
+        return acceptEitherAsync(other, action, ForkJoinPool.commonPool());
+    }
+
+    @Override
+    public FutureCompletionStage<Void> runAfterEither(CompletionStage<?> other, Runnable action) {
+        return runAfterEitherAsync(other, action, SAME_AS_FUTURE);
+    }
+
+    @Override
+    public FutureCompletionStage<Void> runAfterEitherAsync(CompletionStage<?> other, Runnable action) {
+        return runAfterEitherAsync(other, action, ForkJoinPool.commonPool());
+    }
+
+    @Override
+    public <U> FutureCompletionStage<U> thenCompose(Function<? super V, ? extends CompletionStage<U>> fn) {
+        return thenComposeAsync(fn, SAME_AS_FUTURE);
+    }
+
+    @Override
+    public <U> FutureCompletionStage<U> thenComposeAsync(Function<? super V, ? extends CompletionStage<U>> fn) {
+        return thenComposeAsync(fn, ForkJoinPool.commonPool());
+    }
+
+    @Override
+    public FutureCompletionStage<V> whenComplete(BiConsumer<? super V, ? super Throwable> action) {
+        return whenCompleteAsync(action, SAME_AS_FUTURE);
+    }
+
+    @Override
+    public FutureCompletionStage<V> whenCompleteAsync(BiConsumer<? super V, ? super Throwable> action) {
+        return whenCompleteAsync(action, ForkJoinPool.commonPool());
+    }
+
+    @Override
+    public <U> FutureCompletionStage<U> handle(BiFunction<? super V, Throwable, ? extends U> fn) {
+        return handleAsync(fn, SAME_AS_FUTURE);
+    }
+
+    @Override
+    public <U> FutureCompletionStage<U> handleAsync(BiFunction<? super V, Throwable, ? extends U> fn) {
+        return handleAsync(fn, ForkJoinPool.commonPool());
+    }
+
+    @Override
+    public <U> FutureCompletionStage<U> thenApplyAsync(Function<? super V, ? extends U> fn, Executor executor) {
+        requireNonNull(fn, "fn");
+        requireNonNull(executor, "executor");
+
+        Promise<U> promise = executor().newPromise();
+        future.addListener(future -> {
+            Throwable cause = future.cause();
+            if (cause == null) {
+                @SuppressWarnings("unchecked") V value = (V) future.getNow();
+                if (executeDirectly(executor)) {
+                    thenApplyAsync0(promise, value, fn);
+                } else {
+                    safeExecute(executor, () -> thenApplyAsync0(promise, value, fn), promise);
+                }
+            } else {
+                promise.setFailure(cause);
```

Comment:
Not sure if cancel has been killed yet and/or how `isCancelled()` is implemented but it maybe needs to be taken care of as a special case.

... same thing at a few other places further down in this class.

Reply:
good point ... let me add some code